cpp/memilio/mobility/metapopulation_mobility_stochastic.h

```diff
@@ -206,10 +206,12 @@ void MigrationEdgeStochastic::apply_migration(size_t event, SimulationNode<Sim>&
  * edge functor for migration simulation.
  * @see MigrationEdgeStochastic::apply_migration
  */
-template <class Sim, class StochasticEdge>
+template <class Sim, class StochasticEdge,
+          std::enable_if_t<std::is_same<StochasticEdge, MigrationEdgeStochastic>::value, bool> tag = true>
 void apply_migration(StochasticEdge& migrationEdge, size_t event, SimulationNode<Sim>& node_from,
                      SimulationNode<Sim>& node_to)
 {
+    static_assert(tag == true, "tag has to be true");
     migrationEdge.apply_migration(event, node_from, node_to);
 }
```

Comment:
I think the reason for the enable if is that if we delete it we cannot differentiate this apply_migration from the one for non-stochastic edges, because they both then have the same number of templates. Therefore we need the enable_if

Reply:
I'd agree that enable_if might be necessary but why do you need the static assert? If it is false, this function will not be compiled, will it?

Reply:
Yes we can delete that

Reply:
> we cannot differentiate this apply_migration from the one for non-stochastic edges, because they both then have the same number of templates

Sure, we could use this third template to differentiate the functions. But it does not matter *what* that template is. It could really be anything, even a boat. It does not have to be an enable_if.
Usually you use enable_if to resolve functions that look exactly the same otherwise, but the apply_migration functions have different signatures (they have both a different number and types of arguments), so we do not need to do that here.

However, with the enable_if and assert, we essentially have two "useless" templates, since the second template argument is forced to be one specific type. So we could as well only use the Sim template, and replace the Edge template by the type enforced anyways.

I would prefer, if you want to have a templated edge type, that we drop the third template and specify the type of apply_migration in make_migration_sim. Please look at my next comment in this file.
If you want the edge type to be fixed instead, we should drop both the second and third template. This too will resolve the ambiguity here.